src/lib.rs

```diff
@@ -1,16 +1,165 @@
+use std::{
+    collections::BinaryHeap,
+    fmt::Debug,
+    iter::zip,
+    ops::{Add, Deref, Div, Mul, Sub},
+};
+
 use serde::{Deserialize, Serialize};
 
+#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize)]
+#[repr(transparent)]
+pub struct NonNanFloat(f64);
+
+impl PartialEq for NonNanFloat {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        if self.0.is_nan() {
+            other.0.is_nan()
+        } else {
+            self.0 == other.0
+        }
+    }
+}
+
+impl PartialEq<f64> for NonNanFloat {
+    #[inline]
+    fn eq(&self, other: &f64) -> bool {
+        if self.0.is_nan() {
+            other.is_nan()
+        } else {
+            self.0 == *other
+        }
+    }
+}
+
+impl PartialOrd for NonNanFloat {
+    #[inline]
+    #[allow(clippy::all)]
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl Eq for NonNanFloat {}
+
+impl Ord for NonNanFloat {
+    #[inline]
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+impl Sub for NonNanFloat {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        Self(self.0 - other.0)
+    }
+}
+
+impl Sub<f64> for NonNanFloat {
+    type Output = Self;
+
+    fn sub(self, other: f64) -> Self {
+        Self(self.0 - other)
+    }
+}
+
+impl Add for NonNanFloat {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        NonNanFloat(self.0 + other.0)
+    }
+}
+
+impl Add<f64> for NonNanFloat {
+    type Output = Self;
+
+    fn add(self, other: f64) -> Self {
+        NonNanFloat(self.0 + other)
+    }
+}
+
+impl Mul for NonNanFloat {
+    type Output = Self;
+
+    fn mul(self, other: Self) -> Self {
+        NonNanFloat(self.0 * other.0)
+    }
+}
+
+impl Mul<f64> for NonNanFloat {
+    type Output = Self;
+
+    fn mul(self, other: f64) -> Self {
+        NonNanFloat(self.0 * other)
+    }
+}
+
+impl Div for NonNanFloat {
+    type Output = Self;
+
+    fn div(self, other: Self) -> Self {
+        NonNanFloat(self.0 / other.0)
+    }
+}
+
+impl Div<f64> for NonNanFloat {
+    type Output = Self;
+
+    fn div(self, other: f64) -> Self {
+        NonNanFloat(self.0 / other)
+    }
+}
+
+impl Deref for NonNanFloat {
+    type Target = f64;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl NonNanFloat {
+    fn new(value: f64) -> Self {
+        assert!(!value.is_nan());
+        NonNanFloat(value)
+    }
+}
+
+impl From<f64> for NonNanFloat {
+    fn from(value: f64) -> Self {
+        NonNanFloat::new(value)
+    }
+}
+
 #[derive(Debug, Default, Clone, Copy, Serialize, Deserialize)]
 pub struct Coordinate {
-    pub longitude: f64,
-    pub latitude: f64,
+    pub longitude: NonNanFloat,
+    pub latitude: NonNanFloat,
+}
+
+pub trait TCoordianteFloat {
```

Comment:
conversion logic is already implemented for f64. Perhaps, Trait is not necessary?

Reply:
I just wanted to limit the types it can take.